src/warpychain/client.go

```diff
@@ -0,0 +1,146 @@
+	"github.com/warp-contracts/syncer/src/utils/sequencer"
+	sequencer_types "github.com/warp-contracts/syncer/src/utils/sequencer/types"
+	"gorm.io/gorm"
+)
+
+type WarpyChainClient struct {
+	db                     *gorm.DB
+	Log                    *logrus.Entry
+	ethClient              *ethclient.Client
+	writeInteractionClient *sequencer.WriteInteractionClient[WarpyChainInput]
+}
+
+func NewWarpyChainClient(config *config.Config) (self *WarpyChainClient) {
+	self = new(WarpyChainClient)
+	self.Log = logger.NewSublogger("warpychain_client")
+	var err error
+	self.ethClient, err = GetEthClient(self.Log, Avax)
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get ETH client")
+		return
+	}
+
+	return
+}
+
+func (self *WarpyChainClient) WithDB(db *gorm.DB) *WarpyChainClient {
+	self.db = db
+	return self
+}
+
+func (self *WarpyChainClient) WithWriteInteractionClient(writeInteractionClient *sequencer.WriteInteractionClient[WarpyChainInput]) *WarpyChainClient {
+	self.writeInteractionClient = writeInteractionClient
+	return self
+}
+
+func (self *WarpyChainClient) getLastSyncedBlockHeight(ctx context.Context) (lastSyncedHeight int64, err error) {
+	self.Log.Debug("Checking last synced height...")
+
+	err = self.db.WithContext(ctx).
+		Raw(`SELECT finished_block_height
+		FROM sync_state
+		WHERE name = 'WarpyChainer';`).
+		Scan(&lastSyncedHeight).Error
+
+	return
+}
+
+func (self *WarpyChainClient) getCurrentBlockHeight() (currentBlockHeight int64, currentBlockHash string, err error) {
+	header, err := self.ethClient.HeaderByNumber(context.Background(), nil)
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get ETH header")
+		return
+	}
+
+	currentBlockHeight = header.Number.Int64()
+	currentBlockHash = header.TxHash.String()
+
+	self.Log.WithField("height", currentBlockHeight).WithField("hash", currentBlockHash).Debug("Current block height")
+
+	return
+}
+
+func (self *WarpyChainClient) getBlockInfo(blockNumber int64) (blockInfo *types.Block, err error) {
+	blockInfo, err = self.ethClient.BlockByNumber(context.Background(), big.NewInt(blockNumber))
+	if err != nil {
+		self.Log.WithError(err).Error("Could not get ETH header")
+	}
+	return
+}
+
+func (self *WarpyChainClient) updateSyncState(ctx context.Context, blockHeight int64, blockHash string) (err error) {
+	err = self.db.WithContext(ctx).
+		Exec(`UPDATE sync_state
+	SET finished_block_height = ?, finished_block_hash = ?
+	WHERE name = 'WarpyChainer';`, blockHeight, blockHash).
+		Error
+
+	if err != nil {
+		self.Log.WithError(err).Error("Could not update sync_state table")
+		return
+	}
+	return
+}
+
+func (self *WarpyChainClient) checkForRedStoneTx(tx *types.Transaction, ctx context.Context) (isRedStoneTx bool) {
+	isRedStoneTx = false
+	if tx.Hash().Hex() == "0x04659c685df5e87d9bb3755d3943442c25bdf773113fa54cf5427aa38729235e" {
+		encodedString := hex.EncodeToString(tx.Data())
+		if strings.Contains(encodedString, "000002ed57011e0000") {
```

Comment:
I would move this magical value to some const, or maybe even env prop, in case it changes.